### ISSUE TITLE
Passing a concurrent queue object over reference through a function is not working

### ISSUE THREAD

**Issue by bharathraja** [cameron314/concurrentqueue#260]
The data type with which the concurrent queue is initialised is a struct of multiple datatypes. The dequeue is in a separate function running in a separate thread. when the function is called through the thread and the reference is passed through std::ref, it is giving error.

sample code:
`
struct frame_data{
	Mat frame;
	int frame_num;
};

void mask (moodycamel::BlockingConcurrentQueue<frame_data*>& capture){
     frame_data item;
     capture->wait_dequeue(item);
}

main(){
moodycamel::BlockingConcurrentQueue<frame_data> capture;

std::thread(mask, std::ref(capture)).detach();
}`

**Comment by bharathraja** [comment]
error log:

In file included from test_shared_queue.cpp:7:0:
/usr/include/c++/7/thread: In instantiation of ‘struct std::thread::_Invoker<std::tuple<void (*)(moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits>), std::reference_wrapper<moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits> > > >’:
/usr/include/c++/7/thread:127:22:   required from ‘std::thread::thread(_Callable&&, _Args&& ...) [with _Callable = void (&)(moodycamel::BlockingConcurrentQueue<frame_data>); _Args = {std::reference_wrapper<moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits> >}]’
test_shared_queue.cpp:176:41:   required from here
/usr/include/c++/7/thread:240:2: error: no matching function for call to ‘std::thread::_Invoker<std::tuple<void (*)(moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits>), std::reference_wrapper<moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits> > > >::_M_invoke(std::thread::_Invoker<std::tuple<void (*)(moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits>), std::reference_wrapper<moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits> > > >::_Indices)’
  operator()()
  ^~~~~~~~
/usr/include/c++/7/thread:231:4: note: candidate: template<long unsigned int ..._Ind> decltype (std::__invoke((_S_declval<_Ind>)()...)) std::thread::_Invoker<_Tuple>::_M_invoke(std::_Index_tuple<_Ind ...>) [with long unsigned int ..._Ind = {_Ind ...}; _Tuple = std::tuple<void (*)(moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits>), std::reference_wrapper<moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits> > >]
    _M_invoke(_Index_tuple<_Ind...>)
    ^~~~~~~~~
/usr/include/c++/7/thread:231:4: note:   template argument deduction/substitution failed:
/usr/include/c++/7/thread: In substitution of ‘template<long unsigned int ..._Ind> decltype (std::__invoke(_S_declval<_Ind>()...)) std::thread::_Invoker<std::tuple<void (*)(moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits>), std::reference_wrapper<moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits> > > >::_M_invoke<_Ind ...>(std::_Index_tuple<_Ind1 ...>) [with long unsigned int ..._Ind = {0, 1}]’:
/usr/include/c++/7/thread:240:2:   required from ‘struct std::thread::_Invoker<std::tuple<void (*)(moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits>), std::reference_wrapper<moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits> > > >’
/usr/include/c++/7/thread:127:22:   required from ‘std::thread::thread(_Callable&&, _Args&& ...) [with _Callable = void (&)(moodycamel::BlockingConcurrentQueue<frame_data>); _Args = {std::reference_wrapper<moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits> >}]’
test_shared_queue.cpp:176:41:   required from here
/usr/include/c++/7/thread:233:29: error: no matching function for call to ‘__invoke(std::__tuple_element_t<0, std::tuple<void (*)(moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits>), std::reference_wrapper<moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits> > > >, std::__tuple_element_t<1, std::tuple<void (*)(moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits>), std::reference_wrapper<moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits> > > >)’
    -> decltype(std::__invoke(_S_declval<_Ind>()...))
                ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/7/tuple:41:0,
                 from /usr/include/c++/7/bits/unique_ptr.h:37,
                 from /usr/include/c++/7/memory:80,
                 from /usr/include/c++/7/thread:39,
                 from test_shared_queue.cpp:7:
/usr/include/c++/7/bits/invoke.h:89:5: note: candidate: template<class _Callable, class ... _Args> constexpr typename std::__invoke_result<_Functor, _ArgTypes>::type std::__invoke(_Callable&&, _Args&& ...)
     __invoke(_Callable&& __fn, _Args&&... __args)
     ^~~~~~~~
/usr/include/c++/7/bits/invoke.h:89:5: note:   template argument deduction/substitution failed:
/usr/include/c++/7/bits/invoke.h: In substitution of ‘template<class _Callable, class ... _Args> constexpr typename std::__invoke_result<_Functor, _ArgTypes>::type std::__invoke(_Callable&&, _Args&& ...) [with _Callable = void (*)(moodycamel::BlockingConcurrentQueue<frame_data>); _Args = {std::reference_wrapper<moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits> >}]’:
/usr/include/c++/7/thread:233:29:   required by substitution of ‘template<long unsigned int ..._Ind> decltype (std::__invoke(_S_declval<_Ind>()...)) std::thread::_Invoker<std::tuple<void (*)(moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits>), std::reference_wrapper<moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits> > > >::_M_invoke<_Ind ...>(std::_Index_tuple<_Ind1 ...>) [with long unsigned int ..._Ind = {0, 1}]’
/usr/include/c++/7/thread:240:2:   required from ‘struct std::thread::_Invoker<std::tuple<void (*)(moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits>), std::reference_wrapper<moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits> > > >’
/usr/include/c++/7/thread:127:22:   required from ‘std::thread::thread(_Callable&&, _Args&& ...) [with _Callable = void (&)(moodycamel::BlockingConcurrentQueue<frame_data>); _Args = {std::reference_wrapper<moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits> >}]’
test_shared_queue.cpp:176:41:   required from here
/usr/include/c++/7/bits/invoke.h:89:5: error: no type named ‘type’ in ‘struct std::__invoke_result<void (*)(moodycamel::BlockingConcurrentQueue<frame_data>), std::reference_wrapper<moodycamel::BlockingConcurrentQueue<frame_data, moodycamel::ConcurrentQueueDefaultTraits> > >’
Makefile:6: recipe for target 'shared_queue' failed
make: *** [shared_queue] Error 1

**Comment by bharathraja** [comment]
I just passed by simple address reference(&) instead of using std::ref, it is working good. I was converting the code to replace zmq with the concurrent queues, was without thinking used the same syntax.
